### PR TITLE
Pullback of rational functions along strict transforms

### PR DESCRIPTION
**Pull request by HechtiDerLachs**
Needed for the fibration hopping in the Oscar book chapter.

@afkafkafk13 : There is one important change in here: I derived `BlowupMorphism` from `AbsCoveredSchemeMorphism`. I know we talked about this some months ago and you had reasons to decide against this. But I didn't remember these reasons and, at least for me, within the last weeks it became more and more clear that it should really be like this. 

Let me know if you still have objections against this.

### CODE REVIEW

**Comment by HechtiDerLachs** [comment]
I introduced a type for lazy composite maps of covered schemes. These should be useful in many places and I needed them to facilitate the moving around of elements of the function fields.

File: experimental/Schemes/BlowupMorphism.jl [hunk 1]
```diff
@@ -151,10 +161,60 @@ function strict_transform(p::BlowupMorphism, inc::CoveredClosedEmbedding)
...
+    #    ancestor. Since `U_amb` is one of the `affine_charts` of `X`, this will work. 
+    #  - pr_res : U_amb -> V_amb has some codomain such that there exists 
+    #    an ancestor `VV` in the `domain` of `inc_dom_cov` such that 
+    #    inc_dom : VV -> W' has a codomain with `V_amb` as an ancestor. 
+    #  - outside the `complement_equation`s of `U` and `V` the projection 
+    #    was an isomorphism. Then the restriction of `pr_res` to those 
+    #    complements in `U_sub` and `V_sub` also is.
+    k = findfirst(x->has_ancestor(y->y===U_amb, codomain(inc_dom_cov[x])), patches(domain(inc_dom_cov)))
+    U_sub = patches(domain(inc_dom_cov))[k]
+    iso_U_sub = _flatten_open_subscheme(U_sub, domain(inc_dom_cov[U_sub]))
```
Comment (simonbrandhorst) [inline experimental/Schemes/BlowupMorphism.jl]: what does this do and why do you do it?

File: experimental/Schemes/BlowupMorphism.jl [hunk 2]
```diff
@@ -151,10 +161,60 @@ function strict_transform(p::BlowupMorphism, inc::CoveredClosedEmbedding)
...
+    U = domain(p_iso)::PrincipalOpenSubset
+    V = codomain(p_iso)::PrincipalOpenSubset
+    V_amb = ambient_scheme(V)
+    U_amb = ambient_scheme(U)
+
+    # Given all the refinements that potentially had to be done, we have 
+    # to do the following.
+    #  - Find a `patch` `U_sub` of the `domain` of `inc_dom_cov` for which 
+    #    inc_cod : U_sub -> W has a codomain `W` which has `U_amb` as an 
+    #    ancestor. Since `U_amb` is one of the `affine_charts` of `X`, this will work.
```
Comment (simonbrandhorst) [inline experimental/Schemes/BlowupMorphism.jl]: Why is `U_amb` a chart of `X`? It seems to me that it is a chart of Y?
What is the domain and codomain of p_iso? (i.e. does it point in the same direction as p? or the opposite one?)

File: experimental/Schemes/CoveredScheme.jl [hunk 3]
```diff
@@ -568,3 +568,162 @@ function CoveredClosedEmbedding(X::AbsCoveredScheme, I::IdealSheaf;
   return CoveredClosedEmbedding(Z, X, cov_inc, ideal_sheaf=I, check=false)
 end
 
+########################################################################
+# Composite morphism of covered schemes
+########################################################################
+
+@attributes mutable struct CompositeCoveredSchemeMorphism{
```
Comment (simonbrandhorst) [inline experimental/Schemes/CoveredScheme.jl:610]: Needs a little documentation about its scope and purpose.

File: experimental/Schemes/CoveredScheme.jl [hunk 4]
```diff
@@ -568,3 +568,162 @@ function CoveredClosedEmbedding(X::AbsCoveredScheme, I::IdealSheaf;
...
+  if get(io, :supercompact, false)
+    print(io, "Composite morphism")
+  else
+    map_string = "$(domain(f)) -> "
+    for i in 2:length(maps(f))
+      map_string = map_string * "$(domain(map(f)[i])) -> "
+    end
+    map_string = map_string * "$(codomain(map(f)[end]))"
+
+    print(io, "Composition of ", map_string)
```
Comment (simonbrandhorst) [inline experimental/Schemes/CoveredScheme.jl]: No need to print to an intermediate string
```suggestion
    print(io, "Composition of ", map_string)
    print(io, domain(f), "  -> ")
    for i in 2:length(maps(f))
      print(io, domain(map(f)[i]), " -> ")
    end
    print(io, codomain(map(f)[end]))
```

File: experimental/Schemes/CoveredScheme.jl [hunk 5]
```diff
@@ -568,3 +568,162 @@ function CoveredClosedEmbedding(X::AbsCoveredScheme, I::IdealSheaf;
...
+  end
+end
+
+function Base.show(io::IO, ::MIME"text/plain", f::CompositeCoveredSchemeMorphism)
+  io = pretty(io)
+  println(io, "Composite morphism of", Indent())
+  for g in maps(f)
+    println(io, g)
+  end
+end
```
Comment (simonbrandhorst) [inline experimental/Schemes/CoveredScheme.jl]:
```suggestion
  end
print(io, Dedent())
end
```

File: experimental/Schemes/FunctionFields.jl [hunk 6]
```diff
@@ -502,6 +502,19 @@ function is_regular(f::VarietyFunctionFieldElem, W::SpecOpen)
...
+  parent(a) === function_field(X) || error("element does not belong to the correct ring")
+  has_attribute(f, :isomorphism_on_open_subset) || error("need an isomorphism on some open subset")
+  f_res = isomorphism_on_open_subset(f)
+  U = domain(f_res)
+  V = codomain(f_res)
+  aa = a[ambient_scheme(U)]
+  f_res_inv = inverse(f_res)
+  num = pullback(f_res_inv)(numerator(aa))
+  den = pullback(f_res_inv)(denominator(aa))
+  bb = fraction(num)//fraction(den)
```
Comment (simonbrandhorst) [inline experimental/Schemes/FunctionFields.jl]: This relies on caching unique parents. Can we get around it?

File: experimental/Schemes/BlowupMorphism.jl (hunk 1 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/BlowupMorphism.jl]: line 192 is deprecated and superfluous. Did you mean that? In that case: Thanks for catching!

File: experimental/Schemes/FunctionFields.jl (hunk 6 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/FunctionFields.jl]: You mean 516? Yes, indeed.

File: experimental/Schemes/BlowupMorphism.jl (hunk 1 above)
Comment (simonbrandhorst) [reply experimental/Schemes/BlowupMorphism.jl]: Yep

File: experimental/Schemes/FunctionFields.jl (hunk 6 above)
Comment (simonbrandhorst) [reply experimental/Schemes/FunctionFields.jl]: Yep

File: experimental/Schemes/BlowupMorphism.jl [hunk 7]
```diff
@@ -151,10 +161,69 @@ function strict_transform(p::BlowupMorphism, inc::CoveredClosedEmbedding)
...
+    V_amb = ambient_scheme(V)
+    U_amb = ambient_scheme(U)
+
+    # We have the following diagram:
+    #             inc_dom
+    #     Z_trans    ↪     X   ⊃ U_amb ⊃ U
+    #
+    #  pr_res↓             ↓ p           ↓ p_iso
+    #        Z       ↪     Y   ⊃ V_amb ⊃ V
+    #             inc_cod
```
Comment (simonbrandhorst) [inline experimental/Schemes/BlowupMorphism.jl]: The code  (lines 156 - 159) and the docstring suggest the following
```suggestion
    # We have the following diagram:
    #             inc_dom
    #     Z_trans    ↪     Y   ⊃ U_amb ⊃ U
    #
    #  pr_res↓             ↓ p           ↓ p_iso
    #        Z       ↪     X   ⊃ V_amb ⊃ V
    #             inc_cod
```

File: experimental/Schemes/BlowupMorphism.jl [hunk 8]
```diff
@@ -151,10 +161,69 @@ function strict_transform(p::BlowupMorphism, inc::CoveredClosedEmbedding)
...
+    #
+    # Given all the refinements that potentially had to be done, we have 
+    # to do the following.
+    #  1. Find a `patch` `U_sub` of the `domain` of `inc_dom_cov` for which 
+    #    inc_dom : U_sub -> W has a codomain `W` which has `U_amb` as an 
+    #    ancestor. Since `U_amb` is one of the `affine_charts` of `X`, this will work.
```
Comment (simonbrandhorst) [inline experimental/Schemes/BlowupMorphism.jl]:
```suggestion
    #    ancestor. Since `U_amb` is one of the `affine_charts` of `Y`, this will work. 
```

**Review by simonbrandhorst — APPROVED** [review]

**Review by simonbrandhorst — APPROVED** [review]

**Comment by simonbrandhorst** [comment]
Let's pray the CI completes this time.

**Pull request merged**